feat: Adding possibility to output all TableFunction in log/CSV + Solving a bug on PVT CSV export

MelReyCG opened the pull request:
This PR aims to:

1. Add the possibility to write `TableFunction`s in the log or in a CSV file through `logLevel="1"` or `writeCSV="1"` (this feature has been requested by TE users),

2. Allow to create tables without layout specification (added a test to illustrate, the goal is to allow to create table without headers, with only data rows)

3. Slightly refactor the `TableFunction` code to remove any PVT concept from its class,

4. Slightly rewrite the `writeCSV` documentation to make it clearer,

5. Add the table coordinates/values min, max and unit metadata at the top of the table,
![image](https://github.com/user-attachments/assets/7a8afd2b-f41c-411d-9839-a31e005eaeee)

6. For 2D tables, used standard variables to produce narrower output lines (and ensured that the first coordinate is on the X axis),
![image](https://github.com/user-attachments/assets/d8b1664c-5593-43b0-860e-8128405402f1)

7. Unify the CSV generation when a table is too big and move the warning message inside the table, while keep showing its metadata.
![image](https://github.com/user-attachments/assets/7bde1dcc-451d-420d-8675-73d80fb51d54)

----

This PR also solves bugs:
1. on saving a PVT, `SinglePhaseWell` and `CompositionalMultiphaseWell` table to a CSV file when an output folder has not been created yet (lacking `makeDirsForPath()` call),

2. A crash on table generation

3. wrong behaviour of merged cell when they share the same column, and are wider than them:
```
----------------------------------------------------------------------------------------------------------------
|                                                 Lorem Ipsum                                                  |
----------------------------------------------------------------------------------------------------------------
|      A       |      B       |      C       |  X  |      A       |      B       |      C       |      C       |
----------------------------------------------------------------------------------------------------------------
|  0123456789  |  0123456789  |  0123456789  |  X  |  0123456789  |  0123456789  |  0123456789  |  0123456789  |
|  0123456789  |  012345678901234567890123456789  |  X  |  0123456789  |  0123456789  |  0123456789  |  0123456789  |
|  012345678901234567890123456789  |  0123456789  |  X  |  0123456789  |  0123456789  |  0123456789  |  0123456789  |
|  0123456789  |  0123456789  |  0123456789  |  X  |  0123456789  |  0123456789  |  0123456789  |  0123456789  |
|  0123456789  |  0123456789  |  0123456789  |  X  |  01234567890123456789012345678901234567890123456789  |  0123456789  |
|  0123456789  |  0123456789  |  0123456789  |  X  |  0123456789  |  01234567890123456789012345678901234567890123456789  |
|  0123456789  |  0123456789  |  0123456789  |  X  |  0123456789  |  0123456789  |  0123456789  |  0123456789  |
----------------------------------------------------------------------------------------------------------------\
```
which will now give:
```
---------------------------------------------------------------------------------------------------------------------------------
|                                                          Lorem Ipsum                                                          |
---------------------------------------------------------------------------------------------------------------------------------
|       A       |        B         |       C       |  X  |       A       |         B         |        C         |       C       |
---------------------------------------------------------------------------------------------------------------------------------
|   0123456789  |      0123456789  |   0123456789  |  X  |   0123456789  |       0123456789  |      0123456789  |   0123456789  |
|   0123456789  |  012345678901234567890123456789  |  X  |   0123456789  |       0123456789  |      0123456789  |   0123456789  |
|  012345678901234567890123456789  |   0123456789  |  X  |   0123456789  |       0123456789  |      0123456789  |   0123456789  |
|   0123456789  |      0123456789  |   0123456789  |  X  |   0123456789  |       0123456789  |      0123456789  |   0123456789  |
|   0123456789  |      0123456789  |   0123456789  |  X  |  01234567890123456789012345678901234567890123456789  |   0123456789  |
|   0123456789  |      0123456789  |   0123456789  |  X  |   0123456789  |  01234567890123456789012345678901234567890123456789  |
|   0123456789  |      0123456789  |   0123456789  |  X  |   0123456789  |       0123456789  |      0123456789  |   0123456789  |
---------------------------------------------------------------------------------------------------------------------------------
```

----

better column continuity through the horizontal separators:
```
--------------------------------------------------      --------------------------------------------------|
|         Nodes         |          Faces         |      |         Nodes         |          Faces          |
--------------------------------------------------      |-----------------------|-------------------------|
| Local | Ghost | Total | Local  | Ghost| Total  |      | Local | Ghost | Total | Local  | Ghost | Total  |
--------------------------------------------------      |-------|-------|-------|--------|-------|--------|
| 9,201 |   0   | 9,201 | 99,784 |   0  | 99,784 |      | 9,201 |   0   | 9,201 | 99,784 |   0   | 99,784 |
--------------------------------------------------  ->  |-------|-------|-------|--------|-------|--------|
| 9,201 |   0   | 9,201 | 99,784 |   0  | 99,784 |      | 9,201 |   0   | 9,201 | 99,784 |   0   | 99,784 |
| 9,201 |   0   | 9,201 | 99,784 |   0  | 99,784 |      | 9,201 |   0   | 9,201 | 99,784 |   0   | 99,784 |
| 9,201 |   0   | 9,201 | 99,784 |   0  | 99,784 |      | 9,201 |   0   | 9,201 | 99,784 |   0   | 99,784 |
--------------------------------------------------      |-----------------------|-------------------------|
|           1           |            1           |      |           1           |            1            |
|           1           |            1           |      |           1           |            1            |
--------------------------------------------------      ---------------------------------------------------
```
----

Sorry to include such scattered work in the PR, but they are all small changes for tables output in log/csv.

----

Last things left to do to finish this PR for review:
- [x] clearer warning when a table is not displayed + link inside the table
- [x] headers for non-2D CSV columns to be reviewed
- [x] limit table HEIGHT and WIDTH
- [x] display 3D (or more) tables in log
- [x] In depth refactor to solve the merging bug, and have a clearer/faster code, and that does not cheat with constness.
- [x] better column continuity through the separators
- [x] real support for tables without layout specification (added a test to illustrate, the goal is to allow to create table without headers)

MelReyCG committed:
Preparing folder for table outputs (CSVs)
- uniformizing behaviour of TableFunction regarding other outputing components (OutputBase...)

MelReyCG committed:
✨ Added ability to output any TableFunction in log / CSV

MelReyCG committed:
🐛 Properly ensure we only output from rank 0

MelReyCG committed:
📝 clearer writeCSV message

MelReyCG committed:
♻️ refactor a bit table header & code clarity

MelReyCG committed:
🎨 🐛 bug & name fix (removing any pvt concept from TableFunction)

MelReyCG committed:
📝 Clarifying a bit getLogLevel() documentation

MelReyCG committed:
📦 our favorite generated files

MelReyCG commented:
Currently, 
- documentation is not completed, 
- table layout is not optimal.
```
---------------------------------
|        totalRateTable         |
---------------------------------
|  unknown [?]  |  unknown [?]  |
---------------------------------
|            0  |         0.03  |
|    160000000  |            0  |
|    500000000  |            0  |
---------------------------------
```

I'm working on polishing these last points.

MelReyCG committed:
📝 docs update

MelReyCG committed:
🎨 ensure that all TableFunction pointers are properly initialized

MelReyCG committed:
⚰️ dead code

MelReyCG committed:
debug lines

MelReyCG committed:
✨ adding support for multiline table titles + improved code style

MelReyCG committed:
🐛 crash fix for tables that have no content

MelReyCG committed:
🐛🎨 fixing how title length is taken into account + clearer variable names & types

MelReyCG committed:
✨ adding unit / source details in title (and shortened header columns/row by using standard variables) + possibility to change column/row major convention for table2Ds

MelReyCG committed:
💄 using relative paths in table titles

MelReyCG committed:
🎨 uniformize Path.xpp code style

MelReyCG committed:
🔇 removing (alot of) debug logs

MelReyCG committed:
🎨 code style & comments

MelReyCG commented:
I'm also adding a bugfix for issue [3412](https://github.com/GEOS-DEV/GEOS/issues/3412)

MelReyCG committed:
🐛 preparing folder for wells CSV outputs

MelReyCG committed:
Merge remote-tracking branch 'origin/develop' into bugfix/rey/csv-export-fail

MelReyCG committed:
📦 our beloved schema

MelReyCG committed:
💄 CSV tables should keep units as they have no title like in the log

MelReyCG committed:
✨ Adding getVariableSymbol() to units to get a standard variable symbol

MelReyCG committed:
💡 comment on Unit enum for correctness

MelReyCG committed:
💄 fixing layout to be more user-friendly

MelReyCG committed:
🎨 uncrustify

MelReyCG committed:
💄 keep full axis description in CSVs

MelReyCG committed:
💄  better units description

arng40 commented on src/coreComponents/common/format/table/TableFormatter.cpp:
```diff
@@ -577,31 +586,29 @@ void TableTextFormatter::adjustColumnWidth( CellLayoutRows & cells,
-
 }
 
 void TableTextFormatter::outputTitleRow( TableLayout & tableLayout,
```
thanks for the multiline support 🙂

MelReyCG replied:
✌️

MelReyCG committed:
💄  Uniformizing coordinates and units + better 1D layout

MelReyCG committed:
🐛 ultimate compil fix

arng40 commented on src/coreComponents/functions/FunctionBase.cpp:
```diff
@@ -81,6 +82,12 @@ void FunctionBase::setOutputDirectory( string const & dir )
+  {
+    makeDirsForPath( dir );
+  }
+  MpiWrapper::barrier( MPI_COMM_GEOS );
```
is it useful ?

MelReyCG replied:
nope, commented

MelReyCG committed:
🐛 🎨 table log output decision is now done consistently at one unique place

MelReyCG committed:
💩 preparing return of N-D tables in log

MelReyCG committed:
✨ adding a Column::addSubColumns signature to be compatible with Table2Ds

MelReyCG committed:
💄 improved "too big table" warning visibility & separated table description

MelReyCG committed:
🎨 uncrustify

MelReyCG committed:
🧪 adding longer titles to test the "too wide table" bug

MelReyCG committed:
🚧 call computeCellWidth() whenever needed

MelReyCG committed:
🎨 clearer condition

MelReyCG committed:
Squashed commit of the following:

commit 7b8a26a94ecd80a94d8517294c2bf8970d025290
Author: arng40 <arnaud.dudes@capgemini.com>
Date:   Fri Feb 21 11:31:15 2025 +0100

    set better alignment to linear solver table

commit 73b70d3e21da889c6febb0fbc034716b2ab28246
Author: arng40 <arnaud.dudes@capgemini.com>
Date:   Thu Feb 20 14:56:08 2025 +0100

    again unused code

commit d373f0bf304a317b40a0a4cbe77f6e63e3d76a21
Merge: 6d06ac142 fb8336194
Author: arng40 <arnaud.dudes@capgemini.com>
Date:   Thu Feb 20 14:48:04 2025 +0100

    Merge remote-tracking branch 'origin/develop' into refactor/dudes/refactoring-logs-into-dynamic-tables

commit 6d06ac14298ba0666ad903d423d7f208c2cc2f11
Author: arng40 <arnaud.dudes@capgemini.com>
Date:   Thu Feb 20 14:46:15 2025 +0100

    unused code

commit c37c0e05db63f5e05155a73dde255e79910677c2
Author: arng40 <arnaud.dudes@capgemini.com>
Date:   Thu Feb 20 14:31:26 2025 +0100

    xsddd

commit 0071a18195c33789ca3a7eaadfbb56d965aec792
Author: arng40 <arnaud.dudes@capgemini.com>
Date:   Thu Feb 20 14:31:06 2025 +0100

    alignment correction + load balancing table v2 + remove unused code

commit fa792c3f243c3ca42ee2e63f2b989cd14b81bdc3
Author: arng40 <arnaud.dudes@capgemini.com>
Date:   Thu Feb 20 14:30:03 2025 +0100

    set mesh partionner smaller and removed some info if rank size = 1

commit fb83361947bce9910cb2e397dc4638b105291504
Author: Matteo Frigo <43966286+matteofrigo5@users.noreply.github.com>
Date:   Wed Feb 19 12:05:33 2025 -0800

    ci: Add Passing Crack to the integrated tests (#3551)

commit 0bef7f73da5955422b992a02010f0a71f2661b11
Author: arng40 <arnaud.dudes@capgemini.com>
Date:   Wed Feb 19 16:54:22 2025 +0100

    fix alignment when subcolumn added

commit f330bae970e0e2d9a09a5123351f9d76bb3525d7
Author: arng40 <arnaud.dudes@capgemini.com>
Date:   Wed Feb 19 15:38:16 2025 +0100

    v2 load balancing

commit d0488aff873c63902b5d5c5163fdbc178ce411c5
Author: Tom Byer <149726499+tjb-ltk@users.noreply.github.com>
Date:   Tue Feb 18 09:31:38 2025 -0800

    fix:  sync well constraint (#3541)

commit 8b90c2981e8a0b45d853e64942af2588c6ceefbb
Author: arng40 <arnaud.dudes@capgemini.com>
Date:   Tue Feb 18 17:30:19 2025 +0100

    load balancing tabl 1st version

commit 32dd2631172d31710af34c2147e5d22ee64dee4b
Author: arng40 <arnaud.dudes@capgemini.com>
Date:   Tue Feb 18 17:29:49 2025 +0100

    fields table 1st version

commit b41f371dce787152883824bd76084668326381fc
Author: arng40 <arnaud.dudes@capgemini.com>
Date:   Tue Feb 18 17:29:35 2025 +0100

    solubility table 1st version

commit 4c298515d9b0895dba634b5ee59bb9da9e3ccfac
Author: arng40 <arnaud.dudes@capgemini.com>
Date:   Tue Feb 18 17:29:06 2025 +0100

    some comment and ranming on table

commit abcd86b0880c93b47824982db28a939ad22f6391
Author: rasimHZ <43099968+rasimHZ@users.noreply.github.com>
Date:   Mon Feb 17 19:31:13 2025 -0600

    feat: Overall Composition (Z) Formulation (#3443)

commit dd5ba47deab3b0c88cc4407d3ff9b5975fd0f3dc
Author: Matteo Cusini <49037133+CusiniM@users.noreply.github.com>
Date:   Mon Feb 17 13:52:25 2025 -0800

    fix: segFault in ThermoPoromechanics EFEM case (#3552)

commit d32d40e950006806dd3a67de730fccac67ae71ab
Author: Pavel Tomin <paveltomin@users.noreply.github.com>
Date:   Mon Feb 17 13:22:55 2025 -0600

    fix: multiphase contact bugfix, add test case to ats (#3547)

commit 78cd7c425d72779731c479e209913e40534a0f88
Author: Pavel Tomin <paveltomin@users.noreply.github.com>
Date:   Sat Feb 15 18:47:01 2025 -0600

    ci: fix 1d edfm case and add it to ats (#3546)

commit e5f410447a5fa2493ed905536473ecde40963b18
Author: Pavel Tomin <paveltomin@users.noreply.github.com>
Date:   Sat Feb 15 16:59:38 2025 -0600

    refactor: remove SolidMechanicsLagrangianSSLE and rename SolidMechanics_LagrangianFEM to SolidMechanicsLagrangianFEM (#3533)

    Co-authored-by: Randolph Settgast <settgast1@llnl.gov>

commit 27e0ca3bb2899d4781f4aaacde0e5a313e3189a8
Author: Arnaud DUDES <155963334+arng40@users.noreply.github.com>
Date:   Sat Feb 15 21:45:53 2025 +0100

    feat: restore elementCount in SourceFluxStatistics log table (#3540)

commit ec81e8bf749bccea96c047a06fb758f1513d40db
Author: Pavel Tomin <paveltomin@users.noreply.github.com>
Date:   Fri Feb 14 20:53:06 2025 -0600

    ci: remove code owners from .integrated_tests.yaml (#3503)

    * fix: remove code owners from .integrated_tests.yaml

commit 76070c6fd3c19a6db4319aea8a1e76a7c3ca0c8a
Author: MelReyCG <122801580+MelReyCG@users.noreply.github.com>
Date:   Fri Feb 14 22:23:50 2025 +0100

    fix: SegFault when a MPI rank has no data of a sub-region (#3529)

commit bcf36f5b8ad65a3d6d03ed0dc70577ebbbb4722c
Author: Randolph Settgast <settgast1@llnl.gov>
Date:   Thu Feb 13 23:16:14 2025 -0800

    refactor: Remove all instances of array1d<string> (#2968)

    * replace some of the array1d<string> with string_array which is an alias for std::vector<string>

    * add consideration for packing of std::vector<T>

    * generalize vector packing/unpacking

    * add more packing functions for vector<string>

    * replace SortedArray<string> with set<string>

    * replace array1d<MPI_XXXX> with vector

    * update baselines

commit 6769693855ab7cbd185985c1f99c296e395ebc76
Author: Randolph Settgast <settgast1@llnl.gov>
Date:   Thu Feb 13 00:44:34 2025 -0800

    ci: avoid creating files from inside docker container as root (#3542)

    * try running docker as nobody

    * repopulate schema with missing entries

    * rtd fixes

    * update TPL_TAG.

    * add spack generated hostconfigs

    * move ubuntu cuda build to streak2 and disable unit tests

commit 983f60d215f7b9856e9f9430d699ad1854d57817
Author: Pavel Tomin <paveltomin@users.noreply.github.com>
Date:   Wed Feb 12 15:17:42 2025 -0600

    ci: Fix unit tests for Sherlock CPU (#3544)

    * ci: Disable unit tests for Sherlock CPU

    * Use calculated critical volumes

    Co-authored-by: Pavel Tomin <ptls@PW0BMYGN>
    Co-authored-by: dkachuma <dickkachuma@googlemail.com>

commit 116560fab2179abd377b78d0c865c9f42456d4e1
Author: Brian Han <han12@llnl.gov>
Date:   Tue Feb 11 23:57:59 2025 -0800

    ci: Spack docker containers (#3039)

    * Update tag, add PySolver snippet, adjust lai setting

    * Generalize umpire dependency

    * Add LAI interface toggle

    * Remove uberenv submodule and uberenv/spack configuration files from GEOS repository

    * Specify path to ATS python3 executable

    * Update docs

    * Update docs with info about LC script

    * Sherlock changes; adjust hypre/trilinos defaults, make lai setting explicit per job

    * disable sherlock unit tests

MelReyCG committed:
🎨 🐛 few code style & keeping parent alignment properly

MelReyCG committed:
⚰️ removing no-op

MelReyCG commented on src/coreComponents/common/format/table/TableFormatter.cpp:
```diff
@@ -194,16 +194,17 @@ void TableTextFormatter::initalizeTableLayout( TableLayout & tableLayout,
 
-  adjustTableWidth( tableLayout, cellsHeaderLayout, cellsDataLayout, separatorLine, nbEnabledColumn );
+  adjustTableWidth( tableLayout, cellsHeaderLayout, cellsDataLayout, tableTotalWidth, nbEnabledColumn );
 }
```
```
  // 
  tableTotalWidth = tableLayout.getBorderMargin() * 2 + 2;
  for( size_t columnId = 0; columnId < referenceRow.size(); ++columnId )
  {
    tableTotalWidth += referenceRow[columnId].m_cellWidth + size_t( columnId > 0 ? tableLayout.getColumnMargin() : 0 );
    nbEnabledColumn += ( referenceRow[columnId].m_cellType != CellType::Hidden ? 1 : 0 );
  }
```
@arng40 is this right to compute this way? `(TableTextFormatter::initalizeTableGrids())`

arng40 replied:
As long as you take into account all the margin and only the enabled column, sure

MelReyCG replied:
so the comment in getBorderMargin() seems wrong, as for `|  abc  |` I get 2.

MelReyCG committed:
🎨 tableLayout refactor (naming, documentation + one constructor)

MelReyCG committed:
🧪 Adding tests that should run (merged cells behaviour, optimization of space)

MelReyCG committed:
🎨 ⚡️ 🐛 in depth refactor to:
1. resolve a bug with wide merged cells,
2. add a compression pass to optimize table width,
3. attempt to propose a clearer code,
4. optimize formatting,
5. improve constness practices.

MelReyCG committed:
✨ allow hidden cells

MelReyCG committed:
✅ adapt test (whitespaces)

MelReyCG committed:
⚡️ avoiding useless memory alloc/dealloc (benched 40% faster)

MelReyCG committed:
⚡️ CellLayout is now a view on the const data

MelReyCG committed:
🎨 🐛 fix table title containment + naming

MelReyCG committed:
⚰️ remove dead code

MelReyCG committed:
🎨 🐛 Used a more reliable "PreparedTableLayout" type to avoid buggy and / or complicated std::move()

MelReyCG committed:
📝docs

MelReyCG committed:
⚡️ precompute showed column count + maxDepth -> columnLayerCount

MelReyCG committed:
🎨 enforced constness without "cheating" (creating non-const copy where not necessary)

MelReyCG committed:
Merge remote-tracking branch 'origin/develop' into bugfix/rey/csv-export-fail

MelReyCG committed:
✅ re-applied new testTable.variadicTest results (wrong merge choice)

MelReyCG committed:
✅ added linebreak in title in a test

MelReyCG committed:
✅ merged with maxColumnWidth feature, unit tests now all pass

MelReyCG committed:
⚰️ dead code

arng40 commented on src/coreComponents/common/format/table/TableFormatter.cpp:
```diff
@@ -157,549 +137,449 @@ template<>
+          }
+          else
+          { // left side of a cell that have a neightboor
+            tableOutput << string( nbColumnSpaces, cellSpaceChar ); // à faire si on a pas été mergé avec la gauche toute
```
forgotten comment

MelReyCG replied:
removed

arng40 commented on src/coreComponents/common/format/StringUtilities.cpp:
```diff
@@ -205,6 +148,84 @@ template string toMetricPrefixString( unsigned long long int const & );
 template string toMetricPrefixString( double const & );
 
+template< typename STRING_T >
+std::vector< STRING_T > divideLines( size_t & linesWidth, string_view value )
```
move to stringutilities ?

MelReyCG replied:
yes, I think it can be useful for someone else.

arng40 commented on src/coreComponents/common/format/table/TableFormatter.cpp:
```diff
@@ -157,549 +137,449 @@ template<>
-  adjustTableWidth( tableLayout, cellsHeaderLayout, cellsDataLayout, separatorLine, nbEnabledColumn );
-}
+  // after that, we can process the merged cells.
+  stretchRowToMergedCellsWidth( referenceRow, headerCellsLayout, tableLayout, false );
```
why do you call this function with headerCellsLayout, we'll never have a merge cell in headers, will we?  Am I missing something ?

MelReyCG replied:
I think so. For this exemple:
```
----------------------------------------------------------
|        Rank        |             Nodes                 |
|                    |                                   |
----------------------------------------------------------
|                    |    Count local    |  Count ghost  |
|                    |  active elements  |   elemnt n    |
|                    |                   |               |
----------------------------------------------------------
```
... here a view is the internal data of `CellLayoutRows & headerCellsLayout` : 
- `t` being the `m_cellType`,
- `w` being the `m_cellWidth`,
```
----------------------------------------------------------
|       "Rank"       |        ""         |     "Nodes"   |
|  (t=Header, w=4)   |(t=MergeNext, w=0) |(t=Header, w=5)|
----------------------------------------------------------
|        ""          |   "Count local"   | "Count ghost" |
|                    | "active elements" |   "elemnt n"  |
|    (t=Header, w=0) |  (t=Header, w=15) |(t=Header,w=11)|
----------------------------------------------------------
```

Is it clear enough?

MelReyCG replied:
added a note to populateHeaderCellsLayout

arng40 replied:
oh ok I see, thanks for the doc clarification

arng40 commented on src/coreComponents/common/format/table/TableFormatter.cpp:
```diff
@@ -37,37 +37,35 @@ TableFormatter::TableFormatter( TableLayout const & tableLayout ):
   for( const auto & column : m_tableLayout.getColumns())
   {
-    for( const auto & str : column.m_header.m_lines )
+    for( const auto & str : column.m_headerLayout.m_lines )
```
auto const

MelReyCG replied:
switched

MelReyCG committed:
🎨 auto const

MelReyCG committed:
📝 docs

MelReyCG committed:
⚰️ disabled barrier on functions setOutputDirectory()

MelReyCG committed:
📝 forgot some notes of mine

MelReyCG committed:
✨✅ better separators in data, consistent with vertical columns

MelReyCG committed:
✨✅ better separators in headers, consistent with vertical columns, connect with data

MelReyCG committed:
🔇 remove debug line

MelReyCG committed:
✅ removed unnecessary LB to keep the same layout

MelReyCG committed:
💄 avoiding double table generation in log

MelReyCG committed:
💄 replaced double separators by empty row (which looked strange)

MelReyCG committed:
🎨 Adding an alias for column lists for better encapsulation in parameters

MelReyCG committed:
🎨 more standard method name

MelReyCG committed:
✨ TableFunction csv/text output: generalized n-d tables (n>=3)

MelReyCG committed:
Merge branch 'develop' into bugfix/rey/csv-export-fail

MelReyCG committed:
Merge remote-tracking branch 'origin/develop' into bugfix/rey/csv-export-fail

MelReyCG committed:
📦 other schema

MelReyCG committed:
🚧 added no-layout formatter (untested)

MelReyCG committed:
Merge remote-tracking branch 'origin/develop' into bugfix/rey/csv-export-fail

MelReyCG committed:
✨ ✅ support table formatting without layout specification, add a unit-test for it

arng40 commented on src/coreComponents/common/format/table/TableFormatter.cpp:
```diff
@@ -157,553 +141,526 @@ template<>
 
-  RowsCellInput inputDataValues( tableData.getTableDataRows());
-  if( inputDataValues.size() > 0 )
+  populateHeaderCellsLayout( tableLayout, headerCellsLayout, inputDataRowsCount );
+  if( hasColumnLayout )
```
I'm wondering if we can add populateTitleLayout(), i.e extract this code :
```
  // we add the title, as a first row with all cells merged in one containing the title text
  if( hasTitle )
  {
    CellLayout const & inputTitleCell = tableLayout.getTitleLayout();
    CellLayoutRow & titleRow = headerCellsLayout[0];
    // we setup a row of cells merging with the last cell containing the title
    titleRow = {
      std::vector< CellLayout >( headerCellsLayout[1].cells.size(),
                                 CellLayout( CellType::MergeNext ) ),              // cells
      inputTitleCell.m_lines.size() // sublinesCount
    };
    titleRow.cells.back() = inputTitleCell;
  }
```
 this way we can just do 
 
```
if(hasColumnLayout)
    populateHeaderCellsLayout( tableLayout, headerCellsLayout, inputDataRowsCount );
```

and in TableFormatter we can just do :
  ```
   size_t const headerColumnsCount = tableLayout.getLowermostColumnsCount()
  ```

MelReyCG replied:
Nice suggestion, this populating process will more robust and it will be useful later to have them uncoupled, let me know if commit 50e17393cd5907bf27baf19bd04105a4d5e4934b properly follows your recommendation.

MelReyCG committed:
Merge remote-tracking branch 'origin/develop' into bugfix/rey/csv-export-fail

MelReyCG committed:
🔧 support for std::filesystem for CUDA/GCC8.5 compilers

MelReyCG committed:
🦺 rebaseline

MelReyCG committed:
📦 schemaschemaschemaschemaschema

MelReyCG committed:
📝 adding a note for error management

MelReyCG committed:
📝 documenting test goal explanation

MelReyCG committed:
🎨 uncoupled title & header layout populating (we will later be able to insert other content in-between more easily)

MelReyCG committed:
🦺 rebaseline

MelReyCG committed:
🎨 uncrustify

MelReyCG committed:
🐛 fix compilation by removing a unused var, and actually use another one.

MelReyCG committed:
🐛 fix compilation by using the right macro

MelReyCG committed:
🎨 uncrustify + better suggestion

MelReyCG committed:
🎨 uncrustify + better suggestion

MelReyCG committed:
🐛 forgot a side case (when we have no data)

MelReyCG committed:
🦺 rebaseline

MelReyCG committed:
Merge commit '5732c5c0be16ead0b846c302ed06d30988da4a9f' into bugfix/rey/csv-export-fail

MelReyCG committed:
🐛 🚧 fix two side-case bugs + docs on a bug to fix/warn later

arng40 commented on src/coreComponents/common/format/table/TableFormatter.cpp:
```diff
@@ -157,553 +141,533 @@ template<>
+        if( isFirstToMerge )
+        { // we detected a set of cells to merge, let's compute its total width
+          size_t mergedColumnsWidth = currentColumnWidth;
+          for( size_t mergedId = columnId + 1; mergedId < numColumns; mergedId++ )
           {
-            updateCellMaxLength( cellsDataLayout[rowIdx2][idxColumn], maxColumnSize );
-            if( cellsDataLayout[rowIdx2][idxColumn].m_cellType == CellType::Separator )
+            TableLayout::CellLayout const & mergedCell = row.cells[mergedId];
+            mergedColumnsWidth += referenceRow.cells[mergedId].m_cellWidth + flexSpaces[mergedId];
+            mergedColumnsWidth += spaceBetweenColumns;
+            if( mergedCell.m_cellType != CellType::MergeNext || columnId == numColumns - 1 )
```
Found 2 bug in the table : 

- First thing first is when we only have 1 line in TableData and set this way :

```
 TableData tableData;
  tableData.addRow( 1500, CellType::MergeNext, CellType::MergeNext, CellType::MergeNext, CellType::MergeNext, CellType::MergeNext );
```
The table is not being formatted and this has to do with this part because I think oversize is never reassign

MelReyCG replied:
Nice catch!

arng40 commented on src/coreComponents/common/format/table/TableFormatter.cpp:
```diff
@@ -157,553 +141,533 @@ template<>
   }
 
-  updateColumnMaxLength( tableLayout, cellsHeaderLayout, cellsDataLayout );
+  // this reference row will store the displayed number & width of all columns (and will be stretched by data & headers)
```
- When we don't have a tableLayout, the reference row will be the front of TableData, but if we set different values to the first tableData the table output will be badly shaped.

```
  TableData tableData;
  tableData.addRow( "ProductB", CellType::MergeNext, 40, CellType::Hidden );
  tableData.addRow( "ProductA", 12345678, 159812312323123, "ProductName" );
  tableData.addSeparator();
  tableData.addRow( CellType::MergeNext, CellType::MergeNext, CellType::MergeNext, 40)
```

MelReyCG committed:
🐛 ✅ finalized the fix for the side-case bug (when only one merged line) + correct unit test

MelReyCG committed:
🧪 added failing test for the side case when the reference row has merged cell(s)

MelReyCG committed:
🐛 ✅ finalized the fix for when the reference row has merged cell : removed the reference-row concept in favor of a columns-width array

MelReyCG committed:
Merge remote-tracking branch 'origin/develop' into bugfix/rey/csv-export-fail

MelReyCG committed:
🐛 ✅ 2 bug fixes: when parent column only is not visible (now hides children) + it is no longer possible to enter wrong visibility value

MelReyCG committed:
📝 docs

MelReyCG committed:
📝 docs

MelReyCG committed:
🐛 completed the hidden column fix in commit e495521d

MelReyCG committed:
🦺 rebaseline

MelReyCG committed:
🐛 🎨 repaired addSubColumns so it really append + added the "Cell" type to base Column on it and propose a type that stores the setting AND the data (for further works)

MelReyCG committed:
🐛 compil fix

MelReyCG committed:
🐛 compil fix

MelReyCG committed:
📝 🎨 Improve Cell object robustness & encapsulation + docs

MelReyCG committed:
📝 🐛 doxygen region not suported for this usecase :'(

MelReyCG committed:
📝 last lacking docs

MelReyCG committed:
🎨 uncrustify

herve-gross commented:
@wrtobin @cssherman @CusiniM can you review this? Thanks!

CusiniM commented on .integrated_tests.yaml:
```diff
@@ -1,6 +1,6 @@
 baselines:
   bucket: geosx
-  baseline: integratedTests/baseline_integratedTests-pr3396-10875-13b594f
+  baseline: integratedTests/baseline_integratedTests-pr3537-10887-9327e03
```
you ll have to redo this.

MelReyCG replied:
Is it because of future merges on develop? If not I don't understand why.

CusiniM commented on src/coreComponents/common/CMakeLists.txt:
```diff
@@ -138,6 +138,16 @@ blt_add_library( NAME       common
 
+# Older g++ version need a hint to use std::filesystem
+if( CMAKE_CXX_COMPILER_ID STREQUAL "GNU" )
+    if( CMAKE_CXX_COMPILER_VERSION VERSION_LESS 9 )
```
seems an okay workaround but do we need to support versions of gcc older than 9?

MelReyCG replied:
Since Pangea3 is still internally compiled with GCC 8.4, I would say so.

CusiniM commented on src/coreComponents/common/CMakeLists.txt:
```diff
@@ -138,6 +138,16 @@ blt_add_library( NAME       common
+        target_link_libraries( common PRIVATE "stdc++fs" )
+    endif ()
+endif ()
+if ( ENABLE_CUDA )
```
why would enabling `CUDA` imply this?

MelReyCG replied:
Without that, I had an issue with the Pangea 3 CI build (GCC 9.4).
Actually, this code is moved from the `unitTests/meshTests` package, I needed to move it in `common` to add functionnalities to `Path.xpp`.

CusiniM commented on src/coreComponents/common/format/table/TableData.cpp:
```diff
@@ -50,34 +50,37 @@ std::vector< std::vector< TableData::CellData > > const & TableData::getTableDat
+  arraySlice1d< real64 const > columAxisCoordinates = columnMajorInputValues ? dim0AxisCoordinates : dim1AxisCoordinates;
+  integer const nCol = columAxisCoordinates.size();
+  integer const nRow = rowAxisCoordinates.size();
+  // TODO: 1. restore the table non-blocking error system. 2. add this assert 3. add any other error to it.
```
maybe open an issue.

MelReyCG replied:
Arnaud is working on this (and the other TODOs you can find in `common/format/table`) in `arng40/feature/table-errors-listing`, he will post a PR soon.

MelReyCG committed:
➖ removal of gcc<9 compiler option

MelReyCG committed:
🚨 🎨 remove some CUDA warning

MelReyCG committed:
Revert "➖ removal of gcc<9 compiler option"

This reverts commit 977a24d5f5142a29488d439b26368086e15044aa.

MelReyCG committed:
Merge remote-tracking branch 'origin/develop' into bugfix/rey/csv-export-fail

MelReyCG committed:
📦 schema

paveltomin committed:
Merge branch 'develop' into bugfix/rey/csv-export-fail

paveltomin committed:
Update .integrated_tests.yaml

paveltomin committed:
Update BASELINE_NOTES.md
